### PR TITLE
feat: auth and base app structure

### PR DESCRIPTION
**Pull request by roamiiing**

### CODE REVIEW

File: src/_app/use-cases/auth/refresh.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  const refreshToken = useRefreshToken()
+
+  const signOut = useSignOut()
+
+  // creating separate client because it would call itself recursively
+  // TODO: solve this
+  const pokeApi = createFoxyPokeClient()
+
+  return async () => {
+    if (!accessToken.value && !refreshToken.value) {
```
Comment (talkenson) [inline src/_app/use-cases/auth/refresh.ts]: Create `exists` function with signature like exists<T>(target: T | undefined | null): target is T

File: src/_app/use-cases/auth/refresh.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      return
+    }
+
+    if (tokenValidity.value) {
```
Comment (talkenson) [inline src/_app/use-cases/auth/refresh.ts]: Create `exists` function with signature like exists<T>(target: T | undefined | null): target is T

File: src/_app/use-cases/auth/refresh.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      return
+    }
+
+    if (!accessToken.value || !refreshToken.value) {
```
Comment (talkenson) [inline src/_app/use-cases/auth/refresh.ts]: Create `exists` function with signature like exists<T>(target: T | undefined | null): target is T

File: src/middleware/check-authorized.global.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import { useRedirectOnAuth } from '~/_app/use-cases/auth'
+
+export default defineNuxtRouteMiddleware((to, from) => {
+  const redirectOnAuth = useRedirectOnAuth()
+
+  const requirement = to.meta.auth as PageAuthRequirements | undefined
+
+  const accessToken = useAccessToken()
+
+  const isAuthorized = computed(() => !!accessToken.value)
```
Comment (talkenson) [inline src/middleware/check-authorized.global.ts]: Create `exists` function with signature like exists<T>(target: T | undefined | null): target is T
`const isAuthorized = computed(() => exists(accessToken.value))`

File: src/plugins/poke-api.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { createFoxyPokeClient } from '~/_app/infrastructure/foxy-poke-api'
+import { useSignOut } from '../_app/use-cases/auth'
+
+export default defineNuxtPlugin(nuxtApp => {
+  const foxyPokeClient = createFoxyPokeClient()
+  nuxtApp.vueApp.provide('POKE_API', foxyPokeClient)
```
Comment (talkenson) [inline src/plugins/poke-api.ts]: There's a constant for `POKE_API`

File: src/plugins/poke-api.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+  const signOut = useSignOut()
+
+  foxyPokeClient._factory.onError(async error => {
+    if (error.result.reason === 'UNAUTHORIZED') {
```
Comment (talkenson) [inline src/plugins/poke-api.ts]: Should we standardize reasons, or just move only `UNAUTHORIZED` to constants?

File: src/pages/index.vue [hunk 7]
```diff
@@ -1,70 +1,12 @@
 <template>
   <div>
     <UiBaseTypography class="text-center">
-      <h1>Поздравляем!</h1>
+      <h1>Ваш профиль</h1>
 
-      <p>Вы авторизованы и теперь можете творить любую дичь.</p>
+      <p v-if="data?.status === 'rejected'">У вас нет профиля :(</p>
```
Comment (talkenson) [inline src/pages/index.vue]: There's ENUM for `rejected` value

**Review by talkenson — CHANGES_REQUESTED** [review]

File: src/_app/use-cases/profile/my-profile.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import { useAccessToken } from '~~/src/composables/auth/auth-cookies'
+
+export const useMyProfile = () => {
+  const pokeApi = usePokeApi()
+  const accessToken = useAccessToken()
+
+  const { data, refresh } = useAsyncData('profile/my', async () => {
+    return pokeApi.profile.getMyProfile({
+      meta: {
+        accessToken: accessToken.value,
```
Comment (talkenson) [inline src/_app/use-cases/profile/my-profile.ts:10]: In the future, will there be an interceptor here?

File: src/plugins/redirect-on-auth.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+export default defineNuxtPlugin(nuxtApp => {
+  const router = useRouter()
+
+  router.beforeEach((to, from, next) => {
+    const accessToken = getIsomorphicCookie(AuthorizationCookies.AccessToken)(
+      nuxtApp,
+    )
+
+    const isAuthorized = !!accessToken
```
Comment (talkenson) [inline src/plugins/redirect-on-auth.ts]: exists function wanted

File: src/_app/use-cases/profile/my-profile.ts (hunk 8 above)
Comment (roamiiing) [reply src/_app/use-cases/profile/my-profile.ts:10]: `before` will be a middleware with it's own `context` parameter

**Review by talkenson — APPROVED** [review]
Great!

**Pull request merged**